plume-models/src/timeline/mod.rs

```diff
@@ -0,0 +1,839 @@
+use diesel::{self, BoolExpressionMethods, ExpressionMethods, QueryDsl, RunQueryDsl};
+
+use lists::List;
+use posts::Post;
+use schema::{posts, timeline, timeline_definition};
+use std::ops::Deref;
+use {Connection, Error, PlumeRocket, Result};
+
+pub(crate) mod query;
+
+use self::query::{QueryError, TimelineQuery};
+
+pub use self::query::Kind;
+
+#[derive(Clone, Debug, PartialEq, Queryable, Identifiable, AsChangeset)]
+#[table_name = "timeline_definition"]
+pub struct Timeline {
+    pub id: i32,
+    pub user_id: Option<i32>,
+    pub name: String,
+    pub query: String,
+}
+
+#[derive(Default, Insertable)]
+#[table_name = "timeline_definition"]
+pub struct NewTimeline {
+    user_id: Option<i32>,
+    name: String,
+    query: String,
+}
+
+#[derive(Default, Insertable)]
+#[table_name = "timeline"]
+struct TimelineEntry {
+    pub post_id: i32,
+    pub timeline_id: i32,
+}
+
+impl Timeline {
+    insert!(timeline_definition, NewTimeline);
+    get!(timeline_definition);
+
+    pub fn find_by_name(conn: &Connection, user_id: Option<i32>, name: &str) -> Result<Self> {
+        if let Some(user_id) = user_id {
+            timeline_definition::table
+                .filter(timeline_definition::user_id.eq(user_id))
+                .filter(timeline_definition::name.eq(name))
+                .first(conn)
+                .map_err(Error::from)
+        } else {
+            timeline_definition::table
+                .filter(timeline_definition::user_id.is_null())
```

Comment:
this is the only difference between the two expressions. I find it odd that there's no sensible, generic way to solve this.

Reply:
Welcome to the party!
Diesel consider that anything eq None unless we are that explicit

Reply:
I've tried for a few hours to add a `eq_optional` via some trait, but the best I got was "only" 4 compiler error, of which the shortest was 4 lines long. I'm giving up on this

Reply:
kudos for trying.